sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/sastokens/SasTokenImpl.java

```diff
@@ -0,0 +1,498 @@
+                .append("/").append(this.databaseName);
+        }
+
+        if (!this.containerName.isEmpty()) {
+            if (this.databaseName.isEmpty()) {
+                throw new IllegalArgumentException("databaseName");
+            }
+
+            resourcePrefixPath.append(Paths.ROOT).append(Paths.COLLECTIONS_PATH_SEGMENT)
+                .append(Paths.ROOT).append(this.containerName);
+        }
+
+        if (!this.resourceName.isEmpty()) {
+            if (this.containerName.isEmpty()) {
+                throw new IllegalArgumentException("containerName");
+            }
+
+            switch (this.resourceKind) {
+                case ITEM:
+                    resourcePrefixPath.append(Paths.ROOT).append(Paths.DOCUMENTS_PATH_SEGMENT);
+                    break;
+                case STORED_PROCEDURE:
+                    resourcePrefixPath.append(Paths.ROOT).append(Paths.STORED_PROCEDURES_PATH_SEGMENT);
+                    break;
+                case USER_DEFINED_FUNCTION:
+                    resourcePrefixPath.append(Paths.ROOT).append(Paths.USER_DEFINED_FUNCTIONS_PATH_SEGMENT);
+                    break;
+                case TRIGGER:
+                    resourcePrefixPath.append(Paths.ROOT).append(Paths.TRIGGERS_PATH_SEGMENT);
+                    break;
+                default:
+                    throw new IllegalArgumentException("resourceKind");
+            }
+
+            resourcePrefixPath.append(Paths.ROOT).append(this.resourceName);
+        }
+
+        resourcePrefixPath.append(Paths.ROOT);
+        this.resourcePath = resourcePrefixPath.toString();
+
+        StringBuilder partitionRanges = new StringBuilder();
+        if (this.partitionKeyValueRanges != null && !this.partitionKeyValueRanges.isEmpty()) {
+            if (this.resourceKind != CosmosContainerChildResourceKind.ITEM) {
+                throw new IllegalArgumentException("partitionKeyValueRanges");
+            }
+
+            this.partitionKeyValueRanges.forEach(range -> partitionRanges.append(range.encode()).append(","));
+        }
+
+        if (this.expiryTime == null) {
+            this.expiryTime = this.startTime.plus(Duration.ofHours(2));
+        }
+
+        if (this.controlPlaneReaderScope == 0) {
+            this.controlPlaneReaderScope |= SCOPE_CONTAINER_READ.value();
+            this.controlPlaneReaderScope |= SCOPE_CONTAINER_READ_OFFER.value();
+        }
+
+        if (this.dataPlaneReaderScope == 0 && this.dataPlaneWriterScope == 0) {
+            this.dataPlaneReaderScope |= SCOPE_CONTAINER_READ_ALL_ACCESS.value();
+        }
+
+        StringBuilder payload = new StringBuilder(this.user).append("\n")
+            .append(this.userTag).append("\n")
+            .append(this.resourcePath).append("\n")
+            .append(partitionRanges).append("\n")
+            .append(String.format("%X", this.startTime.getEpochSecond())).append("\n")
```

Comment:
generally `String.format(.)` will do string regex which is computationally expensive. hence we should avoid in the per request hot path.

Reply:
This is only used as part of constructing the token's payload which is one time thing.